src/porepy/numerics/fv/mpsa.py

```diff
@@ -138,6 +138,181 @@ def rhs(self, g, bound_stress, bc_val, f):
+
+        rob_faces = self.robin_faces(g)
+
+        bound = data["param"].get_bc(self)
+
+        if bound.bc_type == "scalar":
+            bound.is_dir[rob_faces] = True
+            bound.is_neu[rob_faces] = False
+        elif bound.bc_type == "vectorial":
+            bound.is_dir[:, rob_faces] = True
+            bound.is_neu[:, rob_faces] = False
+        else:
+            raise ValueError("Unknow boundary condition type: " + bound.bc_type)
+        if np.sum(bound.is_dir * bound.is_neu) !=0:
+            raise AssertionError('Found faces that are both dirichlet and neuman')
+
+        # Discretize with normal mpsa
+        self.discretize(g, data, **kwargs)
+        stress, bound_stress = data["stress"], data["bound_stress"]
+
+        # Create A and rhs
+        div = fvutils.vector_divergence(g)
+        a = div * stress
+        b = div * bound_stress
+
+        # we find the matrix indices of the robin faces
+        rob_ind = mcolon.mcolon(
+            g.dim * rob_faces, g.dim * rob_faces + g.dim
+        )
+        # We find the sign of the robin faces.
+        sgn_rob = _sign_matrix(g, rob_faces)
+
+        # The displacement on the robin boundary face are considered unknowns,
+        # so we move them over to the lhs. The rhs now only consists of the
+        # external boundary faces
+        b_rob = b[:, rob_ind]
+        b_external = b.copy()
+        sparse_mat.zero_columns(b_external, rob_ind)
+
+        bound_stress_external = bound_stress.copy().tocsc()
+        sparse_mat.zero_columns(bound_stress_external, rob_ind)
+        # We calculate the stress on the robin faces due to the internal cells
+        # and robin face displacements (the other boundaries are on the rhs)
+        roben_stress = sps.hstack(
+            (
+                sgn_rob * stress[rob_ind, :],
+                (sgn_rob * bound_stress[rob_ind, :])[:, rob_ind]
+            )
+        )
+        # we add the displacement on the robin boundaries
+        alpha = data["param"].get_robin_factor()[rob_faces]
+        alpha = np.ravel([alpha]*g.dim, 'F')
+        roben_disp = sps.hstack(
+            (
+                sps.csr_matrix((rob_ind.size, g.dim * g.num_cells)),
+                sps.diags(alpha, 0),
+            )
+        )
+        roben_condition = roben_stress + roben_disp
+        A = sps.vstack((sps.hstack((a, b_rob)), roben_condition), format="csr")
+
+        # negative sign since we have moved b_external from lhs to rhs
+        d_b = -b_external
+        # sps.csr_matrix((int_b_left.size, g.num_faces * g.dim))
+        d_t = (
+            sps.csr_matrix(
+                (np.ones(rob_ind.size), (np.arange(rob_ind.size), rob_ind)),
+                (rob_ind.size, g.num_faces * g.dim),
+            )
+            -sgn_rob * bound_stress_external[rob_ind]
+        )
+
+        b_matrix = sps.vstack((d_b, d_t), format="csr")
+
+        data["b_e"] = b_matrix
+        data["A_e"] = A
```

Comment:
Does not this conflict with the operator way of thinking about the discretization? Can we derive a 'stress' etc? If not, then a better explanation of the method is necessary.

It may be that what we really need is a more nuanced view of variables internal (or on the boundary) to a grid. We usually think that there is a single variable on all cells / faces / .., but what you have here is a cell centered displacement on all cells, combined with displacement on certain faces. I'm not sure how to include such options in a neat way. Suggestions?

Reply:
This is a implementation of Robin condition on a face level. This implementation does not know about the subfaces. I guess this class can be left out of this pull-request